widget/tree.go

```diff
@@ -0,0 +1,792 @@
+	OnNodeSelected func(uid string, node fyne.CanvasObject)              // Called when the Node with the given Unique ID is selected.
+}
+
+// NewTreeWithStrings creates a new tree with the given string map.
+// Data must contain a mapping for the root, which defaults to empty string ("").
+func NewTreeWithStrings(data map[string][]string) (t *Tree) {
+	t = &Tree{
+		Children: func(uid string) (c []string) {
+			c = data[uid]
+			return
+		},
+		IsBranch: func(uid string) (b bool) {
+			_, b = data[uid]
+			return
+		},
+		CreateNode: func(branch bool) fyne.CanvasObject {
+			return NewLabel("Template Object")
+		},
+		UpdateNode: func(uid string, branch bool, node fyne.CanvasObject) {
+			node.(*Label).SetText(uid)
+		},
+	}
+	t.ExtendBaseWidget(t)
+	return
+}
+
+// NewTreeWithFiles creates a new tree with the given file system URI.
+// TODO contents of a directory are sorted with the given sorter
+func NewTreeWithFiles(root fyne.URI) (t *Tree) {
+	t = &Tree{
+		Root: root.String(),
+		Children: func(uid string) (c []string) {
+			luri, err := storage.ListerForURI(storage.NewURI(uid))
+			if err != nil {
+				fyne.LogError("Unable to get lister for "+uid, err)
+			} else {
+				uris, err := luri.List()
+				if err != nil {
+					fyne.LogError("Unable to list "+luri.String(), err)
+				} else {
+					// TODO sort.Slice(uris, sorter)
+					for _, u := range uris {
+						c = append(c, u.String())
+					}
+				}
+			}
+			return
+		},
+		IsBranch: func(uid string) bool {
+			path := strings.TrimPrefix(uid, "file://")
+			fi, err := os.Lstat(path)
+			if err != nil {
+				fyne.LogError("Unable to stat path "+path, err)
+				return false
+			}
+			return fi.IsDir()
+		},
+		CreateNode: func(branch bool) fyne.CanvasObject {
+			return fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.FileIcon()), NewLabel("Template Object"))
+		},
+	}
+	t.UpdateNode = func(uid string, branch bool, node fyne.CanvasObject) {
+		c := node.(*fyne.Container)
+		i := c.Objects[0].(*Icon)
+		var r fyne.Resource
+		if branch {
+			if t.IsBranchOpen(uid) {
+				// Set open folder icon
+				r = theme.FolderOpenIcon()
+			} else {
+				// Set folder icon
+				r = theme.FolderIcon()
+			}
+		} else {
+			// Set file icon
+			r = theme.FileIcon()
+
+			/* TODO Copied from dialog/fileicon.go - perhaps this should be a utility?
+			var res fyne.Resource
+			switch strings.Split(mime.TypeByExtension(filepath.Extension(uid)), "/")[0] {
+			case "application":
+				res = theme.FileApplicationIcon()
+			case "audio":
+				res = theme.FileAudioIcon()
+			case "image":
+				res = theme.FileImageIcon()
+			case "text":
+				res = theme.FileTextIcon()
+			case "video":
+				res = theme.FileVideoIcon()
+			default:
+				res = theme.FileIcon()
+			}
+			*/
+		}
+		i.SetResource(r)
+		l := c.Objects[1].(*Label)
+		if t.Root == uid {
+			l.SetText(uid)
+		} else {
+			l.SetText(filepath.Base(uid))
```

Comment:
As these are all URIs can you use the URI and get Name() to avoid the dependency on filepath?